srml/scored-pool/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+					.iter()
+					.for_each(|(who, _)| {
+						T::Currency::reserve(&who, T::CandidateDeposit::get())
+							.expect("balance too low");
+					});
+
+				<Module<T, I>>::sort_pool(&mut pool);
+				<Pool<T, I>>::put(&pool);
+				<Module<T, I>>::refresh_members(false);
+
+				let members = <Members<T, I>>::get();
+				T::MembershipInitialized::initialize_members(&members);
+			});
+		})
+	}
+}
+
+decl_event!(
+	pub enum Event<T, I=DefaultInstance> where
+		<T as system::Trait>::AccountId,
+	{
+		/// The given member was removed; see the transaction for who.
+		MemberRemoved,
+		/// An entity has issued a candidacy.
+		CandidateAdded,
+		/// An entity withdrew candidacy.
+		CandidateWithdrew,
+		/// The candidacy was forcefully removed for an entity.
+		CandidateKicked,
+		/// Phantom member, never used.
+		Dummy(sr_std::marker::PhantomData<(AccountId, I)>),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait<I>, I: Instance=DefaultInstance>
+		for enum Call
+		where origin: T::Origin
+	{
+		fn deposit_event<T, I>() = default;
+
+		/// Every `Period` blocks the `Members` set is refreshed from the
+		/// highest scoring members in the pool.
+		fn on_initialize(n: T::BlockNumber) {
+			if n % T::Period::get() == Zero::zero() {
+				<Module<T, I>>::refresh_members(true);
+			}
+		}
+
+		/// Add `origin` to the pool of candidates.
+		pub fn issue_candidacy(origin) {
+			let who = ensure_signed(origin)?;
+
+			let _ = Self::find_in_pool(&who)
+				.ok()
+				.map_or_else(
+					|| Ok(()),
+					|_| Err("already a member"),
+				)?;
+
+			T::Currency::reserve(&who, T::CandidateDeposit::get())
+				.map_err(|_| "balance too low")?;
+
+			let mut pool = <Pool<T, I>>::get();
+			pool.push((who.clone(), None));
+			Self::sort_pool(&mut pool);
+			<Pool<T, I>>::put(&pool);
+
+			Self::deposit_event(RawEvent::CandidateAdded);
+		}
+
+		/// An entity withdraws candidacy and gets its deposit back.
+		///
+		/// If the entity is part of the `Members`, then the highest member
+		/// of the `Pool` that is not currently in `Members` is immediately
+		/// placed in the set instead.
+		pub fn withdraw_candidacy(origin) {
+			let who = ensure_signed(origin)?;
+
+			Self::remove_member(who)?;
+			Self::deposit_event(RawEvent::CandidateWithdrew);
+		}
+
+		/// Kick a member `who` from the set.
+		///
+		/// May only be called from `KickOrigin` or root.
+		pub fn kick(origin, who: T::AccountId) {
+			T::KickOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(ensure_root)
+				.map_err(|_| "bad origin")?;
+
+			Self::remove_member(who)?;
+			Self::deposit_event(RawEvent::CandidateKicked);
+		}
+
+		/// Score a member `who` with `score`.
+		///
+		/// May only be called from `ScoreOrigin` or root.
+		pub fn score(origin, who: T::AccountId, score: T::Score) {
+			T::ScoreOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(ensure_root)
+				.map_err(|_| "bad origin")?;
+
+			let mut pool = <Pool<T, I>>::get();
+			let location = Self::find_in_pool(&who)?;
```

Comment:
Since this pool guy is sorted: 
- you do: search (`O(LogN)`) remove and naively push (`O(1)`) and sort (`O(NlogN)`) => `O(NlogN)`
- could do: search and remove and **search again for the new spot** and `insert` => `O(2LogN)`

(didn't think it through thoroughly but seems fair to me)

Reply:
Hey Kian, great spotting! I adapted the code.

Reply:
actually, `Vec::remove` is O(N), so it's O(2LogN + N) =~ O(N). still better than O(N.logN) though, but timsort is fiendishly quick, so there's probably not much in it.